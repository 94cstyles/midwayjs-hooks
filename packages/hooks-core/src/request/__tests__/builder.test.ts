import { FileRouter, ProjectConfig } from '../../'
import {
  createApiClientMatcher,
  K_MATCHER_CALLBACK,
  useApiClientMatcher,
  buildApiClient,
} from '../builder'

describe('api client builder', () => {
  test('create ApiClientMatcher', () => {
    const matcher = createApiClientMatcher()
    expect(matcher).toBeTruthy()
  })

  const testMeta = {
    file: '/index.test',
    functionName: 'test',
    functionId: 'test',
    route: {
      baseDir: '/',
      basePath: '/api',
    },
  }

  test('create ApiClientMatcher with route', () => {
    const matcher = createApiClientMatcher().route(
      {
        basePath: '/api',
      },
      {
        client: 'http-request',
        metadata: {
          url: '/api/test',
        },
      }
    )

    expect(matcher[K_MATCHER_CALLBACK].length).toBe(1)

    const callback = matcher[K_MATCHER_CALLBACK][0]
    expect(callback(testMeta)).toMatchSnapshot()
  })

  test('create ApiClientMatcher with match', () => {
    const matcher = createApiClientMatcher().match((meta) => {
      if (meta.functionName === 'test') {
        return {
          client: 'http-request',
          metadata: {
            url: '/api/functionName',
            method: 'get',
          },
        }
      }
    })

    expect(matcher[K_MATCHER_CALLBACK].length).toBe(1)

    const callback = matcher[K_MATCHER_CALLBACK][0]
    expect(callback(testMeta)).toMatchSnapshot()
  })

  test('use ApiClientMatcher to build api client', async () => {
    const matcher = createApiClientMatcher().route(
      {
        basePath: '/api',
      },
      {
        client: 'http-request',
        metadata: {
          url: '/api/test',
        },
      }
    )

    useApiClientMatcher(matcher)

    const projectConfig: ProjectConfig = {
      source: '/',
      routes: [{ baseDir: 'src', basePath: '/api' }],
    }

    const router = new FileRouter({
      root: '/',
      projectConfig,
      useSourceFile: true,
    })

    const apiClient = await buildApiClient(
      '/src/index.ts',
      `export default () => {}`,
      router,
      false
    )

    expect(apiClient).toMatchSnapshot()
  })

  test('use ApiClientMatcher with multi matcher', async () => {
    const projectConfig: ProjectConfig = {
      source: '/',
      routes: [
        { baseDir: 'src', basePath: '/about' },
        { baseDir: 'event', event: 'wechat-miniprogram' },
      ],
    }

    const router = new FileRouter({
      root: '/',
      projectConfig,
      useSourceFile: true,
    })

    const http = createApiClientMatcher().route(
      {
        basePath: '/about',
      },
      {
        client: 'http-request',
        metadata: {
          url: '/about',
        },
      }
    )

    const event = createApiClientMatcher().route(
      {
        event: 'wechat-miniprogram',
      },
      {
        client: 'hooks-miniprogram-client',
      }
    )

    useApiClientMatcher(http)
    useApiClientMatcher(event)

    const httpClient = await buildApiClient(
      '/src/index.ts',
      `export default () => {}`,
      router,
      false
    )

    const eventClient = await buildApiClient(
      '/event/index.ts',
      'export const getOpenId = () => return 123456',
      router,
      false
    )

    expect(httpClient).toMatchSnapshot()
    expect(eventClient).toMatchSnapshot()
  })
})
