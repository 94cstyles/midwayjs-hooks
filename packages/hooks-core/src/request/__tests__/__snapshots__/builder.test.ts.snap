// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api client builder create ApiClientMatcher with match 1`] = `
Object {
  "client": "http-request",
  "metadata": Object {
    "method": "get",
    "url": "/api/functionName",
  },
}
`;

exports[`api client builder create ApiClientMatcher with route 1`] = `
Object {
  "client": "http-request",
  "metadata": Object {
    "url": "/api/test",
  },
}
`;

exports[`api client builder use ApiClientMatcher to build api client 1`] = `
"import { request as request$0 } from 'http-request';

export default function $default(...args) {
  const options = Object.assign(
    {
      functionId: 'src-index',
      data: { args },
    },
    {
      url: '/api/test',
    }
  );

  return request$0(options);
}
"
`;

exports[`api client builder use ApiClientMatcher with multi matcher 1`] = `
"import { request as request$0 } from 'http-request';

export default function $default(...args) {
  const options = Object.assign(
    {
      functionId: 'src-index',
      data: { args },
    },
    {
      url: '/about',
    }
  );

  return request$0(options);
}
"
`;

exports[`api client builder use ApiClientMatcher with multi matcher 2`] = `
"import { request as request$0 } from 'hooks-miniprogram-client';

export function getOpenId(...args) {
  const options = Object.assign(
    {
      functionId: 'event-index-getopenid',
      data: { args },
    },
    {}
  );

  return request$0(options);
}
"
`;
